programmers-coding-test/minjae/dfs-bfs/BFS-예제.js

```diff
@@ -0,0 +1,30 @@
+const graph = {
+    1: [2, 3, 8],
+    2: [1, 7],
+    3: [1, 4, 5],
+    4: [3, 5],
+    5: [3, 4],
+    6: [7],
+    7: [2, 6, 8],
+    8: [1, 7],
+};
+
+//BFS는 큐 2개 사용
+function bfs(graph, startNode) {
+    let visitedQueue, needVisitQueue; 
+    visitedQueue = needVisitQueue = [];
+
+    needVisitQueue.push(startNode);
+
+    while (needVisitQueue.length !== 0) {
+        const node = needVisitQueue.shift(); //선입 선출
+        
+        if (!visitedQueue.includes(node)) {
+            visitedQueue.push(node);
+            needVisitQueue = [...needVisitQueue, ...graph[node]];
+        }
+    }
```

Comment:
리스트를 초기화할 때 기존에 처럼
```javascript
let a, b;
a = b = [];
```
라고 하게 되면 a와 b는 같은 메모리 공간을 참조하게 됩니다.
따라서  **a.push(1)** 을 할 때에 **b**에도 **1**이 추가가 되는 불상사가 일어나게 돼요.
때문에 배열이나 오브젝트 같이 객체의 내부 메모리를 참조해야하는 변수를 선언할 때는 **a = b = []**는 안하는게 좋습니다.
그냥 아래처럼 구조분해할당으로 객체를 초기화하는 식으로 코딩 습관을 바꾸시는게 좋을 것 같아요.
제가 이상한 걸 알려드려가꼬 괜히 ... 죄송합니다 ㅜㅜ

또 배열이나 리스트를 직접적으로 교체하는 것은 안좋습니다.
```javascript
let a = [];
a = [1,2];
a = 4;
a = "123";
```
등과 같이 말이죠.
const로 선언해서 해당 배열은 배열로만 사용하고 데이터만 추가하는 방향으로 가는 것이 데이터 일관성을 지키는 방법입니다.
```javascript
const a = [];
a.push(1);
```
```suggestion
   const [visitedQueue, needVisitQueue] = [[], []];

    needVisitQueue.push(startNode);

    while (needVisitQueue.length !== 0) {
        const node = needVisitQueue.shift(); //선입 선출
        
        if (!visitedQueue.includes(node)) {
            visitedQueue.push(node);
            needVisitQueue.push(...graph[node]);
        }
    }
```